lib/rspec/mocks/method_double.rb

```diff
@@ -215,72 +218,39 @@ def raise_method_not_stubbed_error
-            # we must create a new prepend module, even if one exists later,
-            # because ours will only take precedence if it comes first.
-            return new_rspec_prepended_module if mod.method_defined?(method_name)
-          end
+      # We subclass `Module` in order to be able to easily detect our prepended module.
+      RSpecPrependedModule = Class.new(Module)
 
-          nil
-        end
+      def definition_target
+        @definition_target ||= usable_rspec_prepended_module || object_singleton_class
+      end
 
-        def new_rspec_prepended_module
-          RSpecPrependedModule.new.tap do |mod|
-            object_singleton_class.__send__ :prepend, mod
-          end
+      def usable_rspec_prepended_module
+        @proxy.prepended_modules_of_singleton_class.each do |mod|
+          # If we have one of our modules prepended before one of the user's
+          # modules that defines the method, use that, since our module's
+          # definition will take precedence.
+          return mod if RSpecPrependedModule === mod
+
+          # If we hit a user module with the method defined first,
+          # we must create a new prepend module, even if one exists later,
+          # because ours will only take precedence if it comes first.
+          return new_rspec_prepended_module if mod.method_defined?(method_name)
         end
 
-      else
-
-        private
+        nil
+      end
 
-        def definition_target
-          object_singleton_class
+      def new_rspec_prepended_module
+        RSpecPrependedModule.new.tap do |mod|
+          @object.singleton_class.prepend mod
         end
-
       end
 
-    private
-
       def remove_method_from_definition_target
+        # In Ruby 2.4 and earlier, `remove_method` is private
         definition_target.__send__(:remove_method, @method_name)
-      rescue NameError
-        # This can happen when the method has been monkeyed with by
-        # something outside RSpec. This happens, for example, when
-        # `file.write` has been stubbed, and then `file.reopen(other_io)`
-        # is later called, as `File#reopen` appears to redefine `write`.
-        #
-        # Note: we could avoid rescuing this by checking
-        # `definition_target.instance_method(@method_name).owner == definition_target`,
-        # saving us from the cost of the expensive exception, but this error is
-        # extremely rare (it was discovered on 2014-12-30, only happens on
-        # RUBY_VERSION < 2.0 and our spec suite only hits this condition once),
-        # so we'd rather avoid the cost of that check for every method double,
-        # and risk the rare situation where this exception will get raised.
-        RSpec.warn_with(
-          "WARNING: RSpec could not fully restore #{@object.inspect}." \
-          "#{@method_name}, possibly because the method has been redefined " \
-          "by something outside of RSpec."
-        )
```

Comment:
Restore this and amend the comment? I still think it is a valid protection even if the specific example provided doesn't happen anymore?

Reply:
Strongly agree 👍
A tricky situation that is out of our control.